product_print_category/models/product_template.py

```diff
@@ -18,33 +18,14 @@ class ProductTemplate(models.Model):
 
     to_print = fields.Boolean(related="product_variant_id.to_print", readonly=False)
 
-    @api.depends("product_variant_ids")
-    def _compute_product_variant_id(self):
-        # This is a very particular workaround for the following workflow:
-        #
-        # - Archive product.template.
-        # - Duplicate product.template. <-- This method is called in this step.
-        # - Unarchive duplicated product.template.
-        #
-        # If a product.template is archived, then its variants are also
-        # archived. This means that accessing `product_variant_ids` outputs an
-        # empty recordset, subsequently causing this compute function to wrongly
-        # populate `product_variant_id` with null. By disabling `active_test`,
-        # we prevent this problem.
-        #
-        # Upsteam bugfix in <https://github.com/odoo/odoo/pull/181811> targeted
-        # at v15+.
-        for p in self:
-            # We do with_context() on each individual product instead of on
-            # self, because doing it on self does not produce the desired result
-            # in Odoo 12, somehow.
-            p.product_variant_id = (
-                p.with_context(active_test=False).product_variant_ids[:1].id
-            )
-
     @api.multi
     def write(self, vals):
         res = super().write(vals)
         if self.env.context.get("update_to_print_category", True):
             self._update_to_print_values(vals)
+        # Recompute product_variant_id to circumvent a bug where
+        # product_variant_ids cannot be accessed normally (empty recordset)
+        # because they are archived at time of computation.
+        if "active" in vals:
+            self._compute_product_variant_id()
```

Comment:
Fixed the test.

Between trying to upstream this and patching it here, I've also had a wrong assumption about how things work in Odoo 12.

In Odoo 12:

- When copying a product.template which has product variants, the copied result does NOT have any product variants.

In Odoo 15 (and maybe earlier):

- When copying a product.template which has product variants, the `product.template.attribute.value`s (PTAVs) from which those `product.product`s were generated are copied. Then, during the `create` method of the product.template, new `product.product`s are created from those PTAVs.

---

To answer your question:

Indeed, when a product.template is inactive, `_compute_product_variant_id()` sets False, because `product_variant_ids` only contains archived products and will thus be an empty recordset. This appears to be by design, so we match upstream behaviour here.

When a template is (re)activated, instead of reactivating its product variants, **new** `product.product`s are created from the PTAVs. Then by running our compute method after that has happened, `product_variant_id` is correctly populated.

But by running that compute method again when the template (and thus its also its variants) is unarchived, we get a sensible result. This matches upstream behaviour precisely.

Reply:
nicely done, thanks for the explanation!